pkg/internal/vault/vault_test.go

```diff
@@ -146,29 +233,45 @@ func TestSign(t *testing.T) {
 		"a good csr and good response should return a certificate": {
 			csrPEM: csrPEM,
 			issuer: gen.Issuer("vault-issuer",
-				gen.SetIssuerVault(v1.VaultIssuer{}),
+				gen.SetIssuerVault(cmapi.VaultIssuer{}),
 			),
 			fakeClient: vaultfake.NewFakeClient().WithRawRequest(&vault.Response{
 				Response: &http.Response{
 					Body: ioutil.NopCloser(bytes.NewReader(bundleData))},
 			}, nil),
 			expectedErr:  nil,
 			expectedCert: testCertBundle,
-			expectedCA:   testCertBundle,
+			expectedCA:   testIntermediateCa,
 		},
 
 		"vault issuer with namespace specified": {
 			csrPEM: csrPEM,
 			issuer: gen.Issuer("vault-issuer",
-				gen.SetIssuerVault(v1.VaultIssuer{Namespace: "test"}),
+				gen.SetIssuerVault(cmapi.VaultIssuer{Namespace: "test"}),
 			),
 			fakeClient: vaultfake.NewFakeClient().WithRawRequest(&vault.Response{
 				Response: &http.Response{
 					Body: ioutil.NopCloser(bytes.NewReader(bundleData))},
 			}, nil),
 			expectedErr:  nil,
 			expectedCert: testCertBundle,
-			expectedCA:   testCertBundle,
+			expectedCA:   testIntermediateCa,
+		},
+
+		"vault issuer with multiple CA certificates": {
+			csrPEM: csrPEM,
+			issuer: gen.Issuer("vault-issuer",
+				gen.SetIssuerVault(cmapi.VaultIssuer{
+					Namespace: "test",
+				}),
+			),
+			fakeClient: vaultfake.NewFakeClient().WithRawRequest(&vault.Response{
+				Response: &http.Response{
+					Body: ioutil.NopCloser(bytes.NewReader(multiCABundle))},
+			}, nil),
+			expectedErr:  nil,
+			expectedCert: testCertBundle,
```

Comment:
just to enforce what i mentioned on the [PR comment](https://github.com/jetstack/cert-manager/pull/3340#issuecomment-702276174) this code test only the leaf certificate (not PEM bundle) as expected certificate.

ps.: would really appreciate if the change i've proposed enters on this PR. i've tested by rearranging manually to:
ca.crt => with certificate chain (...intermediateN CA + root CA)
tls.crt => with leaf certificate
tls.key => with leaf key

mutual TLS worked fine in Istio when configured like this.

Reply:
I agree this seem the most correct implementation, but I'm weary because:
- I think it may be valid for the CA certificates to be in `tls.crt` -- I think some server applications will use them if present, and not use a `ca.crt` file
- Some users may be depending on this feature

Does mutual TLS fail without it structured as you recommend?

Reply:
Usually the intermediate CAs are in `tls.crt` as the server has to send these along in the handshake

Reply:
Altough it seems hacky, it's true not all (maybe most) of solutions don't use `ca.crt`, maybe because it's not provided when using a globally trusted CA.

I guess the right process would be that solutions would use both and create a bundle with them when necessary.

I just hope it works well with Istio, with certificates being repeated. 🤞

Reply:
I would expect that the `ca.crt` field contains the single root CA certificate.

The `tls.crt` field would then contain the certificate bundle, including the intermediary and leaf certificate.

Reply:
And why would you expect that?

I would expect `tls.crt` and `tls.key`, as the keypair of the leaf certificate, and any CA related certificates would be in `ca.crt`. Just feels more natural to me to have the keypair and the CA certificate (chain or not) on another file.

Eg.: https://kubernetes.github.io/ingress-nginx/examples/auth/client-certs/#client-certificate-authentication

>1. CA certificate and Key(Intermediate Certs need to be in CA)
>2. Server Certificate(Signed by CA) and Key (CN should be equal the hostname you will use)
>3. Client Certificate(Signed by CA) and Key

"Intermediate Certs need to be in CA"

I can search for other examples if you'd like to, but i've been struggling and learning on certificate management on past few weeks to endure security on systems i'm working on and i'm pretty confident of what i'm saying.

Reply:
It would surely bring challanges to solutions using only `tls.crt` as the bundle but it doesn't make the current approach right. Maybe we could provide another key to the secret: `bundle.pem` (tls.crt + ca.crt)

Reply:
The reason it's done in this way is that typically, root CA certificates are installed into user's browsers/operating systems. It is then up to the thing utilising the leaf certificate to prove it is trusted all the way up to that trusted root. You would not want to store the intermediate certificates in the clients trust store typically (take a look in ca-certificates and you'll see that this is the case).

We don't have a strict definition today for what `ca.crt` *should* contain, but I certainly do think we should include the chain in the `tls.crt` key (and besides, we can't break this behaviour now anyway as people will rely on it, and we will break existing configurations)

Reply:
I'm all up for the non-breaking changes for now, no questions on that. But the current approach is no the clearest since:
1. cert-manager is already relying on `ca.crt`, so maybe we could change the secret type to Opaque and drop the argument that Kubernetes TLS secret type doesn't even support it officially.
2. ACME and webservers are not the only use-cases, so at least others should be considered also, maybe a configuration that toggles behaviors, i don't know, but the current it's too strict to the current flow, which is not agnostic, as i thought cert-manager should be.
3. currently, the issuer CA cert is being repeated on both `ca.crt` and `tls.crt`, and it gets worse in Vault's specific case after this PR is merged, since it repeates the whole CA chain, not only the issuer CA cert, on both, which surely shows something is off, at least to me,

Again, what you fellas are mentioning as the certificate (tls.crt), is actually the bundle, so it would be more clear to have a `bundle.pem` or something of a sort, and leave the `tls.crt` and `ca.crt` only with the leaf certificate and CA chain certificates respectively.

Reply:
So.. for the avoidance of doubt, the intention when `ca.crt` was first introduce in cert-manager, and the intention when `tls.crt` was first added to (i.e. day zero of the project) was:

`tls.crt` - contains the leaf certificate and all intermediates leading up to the root CA (but *not* including the root CA)
`ca.crt` - contains the root CA

Therefore, if `tls.crt` is presented, the only certificate that needs to be in the client's trust store is `ca.crt`.

Reply:
Totally agreed that the current behaviour with Vault is incorrect, and I also suspect some other issuer types may mis-use that `ca.crt` field too (hence the comment that we've been too lax about our definition of what `ca.crt` should contain).

The other difficult part about this is that it sometimes is not possible to actually retrieve the `ca.crt` for a given Issuer implementation (i.e. ACME as a *protocol* does not actually expose this). In those cases, `ca.crt` should be *empty*.

In terms of longer term solutions for trust distribution, I'm looking at https://github.com/kubernetes/kubernetes/issues/63726 to see if we can adopt a similar pattern and have *Issuer* implementations actually populate a TrustBundle resource which can then be consumed by Pods/whatever else needs to verify trust in future.

Reply:
> It would surely bring challanges to solutions using only `tls.crt` as the bundle but it doesn't make the current approach right. Maybe we could provide another key to the secret: `bundle.pem` (tls.crt + ca.crt)

This looks similar: https://github.com/jetstack/cert-manager/issues/843

Reply:
Perhaps there is some distinction between use-cases here, as I am looking to create certificates for mTLS. That's driving some different behavior in clients and the certificate files that they expect.

Reply:
@crandles what worked for me, for Istio mTLS, was to ensure `ca.crt` was filled with the full chain (single or multi certs), as for the `tls.crt` there was no difference by having the whole bundle (leaf keypair + cert chain) or just the leaf certificate, both worked fine after all.

Your approach on code will work just fine, at least in my specific case.